src/compiler/checker.ts

```diff
@@ -32423,22 +32423,23 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getContextualCallSignature(type, node);
         }
         let signatureList: Signature[] | undefined;
-        const types = (type as UnionType).types;
+        const types = (type as UnionType).types
+            .map(type => ({ type, signature: getContextualCallSignature(type, node) }))
+            .filter(type => type.signature)
+            .sort((t1, t2) => t1.signature!.parameters.length - t2.signature!.parameters.length);
         for (const current of types) {
-            const signature = getContextualCallSignature(current, node);
-            if (signature) {
-                if (!signatureList) {
-                    // This signature will contribute to contextual union signature
-                    signatureList = [signature];
-                }
-                else if (!compareSignaturesIdentical(signatureList[0], signature, /*partialMatch*/ false, /*ignoreThisTypes*/ true, /*ignoreReturnTypes*/ true, compareTypesIdentical)) {
-                    // Signatures aren't identical, do not use
-                    return undefined;
-                }
-                else {
-                    // Use this signature for contextual union signature
-                    signatureList.push(signature);
-                }
+            const signature = current.signature!;
+            if (!signatureList) {
+                // This signature will contribute to contextual union signature
+                signatureList = [signature];
+            }
+            else if (!compareSignaturesIdentical(signatureList[0], signature, /*partialMatch*/ true, /*ignoreThisTypes*/ true, /*ignoreReturnTypes*/ true, compareTypesSubtypeOf)) {

```

Comment:
If you are loosening here how parameter types are meant to be compared I'd probably expect here:
```suggestion
            else if (!compareSignaturesIdentical(signatureList[0], signature, /*partialMatch*/ true, /*ignoreThisTypes*/ true, /*ignoreReturnTypes*/ true, compareTypesAssignable)) {
```

Unless there is already a reason why the subtype relationship was used here.

Reply:
I'm not really sure when each relation should be used. I used `compareTypesSubtypeOf` because that's what is used in a call to `compareSignaturesIdentical` with `partialMatch === true` in the function `findMatchingSignatures`.

Reply:
`findMatchingSignatures` gets only called by `getUnionSignatures` so I agree that this probably makes sense